Documentation/designs/prometheus-operator-cli.md

```diff
@@ -0,0 +1,61 @@
+
+### Why my Prometheus resource is not being created?
+
+People were struggling to troubleshoot why their Prometheus resources were not being created. They were not sure if the issue was with the Prometheus Operator or with the Prometheus resource itself.
+After long investigation, they found out that the Prometheus Operator were only watching the namespace where the Prometheus Operator was deployed, and the Prometheus resource was being created in a different namespace.
+
+### Why my ServiceMonitor is not being created?
+
+People were struggling to troubleshoot why their ServiceMonitor resources were not being created. They were not sure if the issue was with the Prometheus Operator or with the ServiceMonitor resource itself.
+
+After long investigation, they found out that the matchingLabels field in the ServiceMonitor resource was not matching the labels of the target Service.
+
+These are just some common issues that users reported, but there are many other issues that users face when managing Prometheus Operator resources, and we believe that a CLI tool can help users to manage Prometheus Operator resources more easily and efficiently.
+
+## Proposal
+
+We propose to create a new CLI tool to manage Prometheus Operator resources, allowing users to create, update, and delete Prometheus Operator resources using a simple CLI interface as well as troubleshoot and debug Prometheus Operator resources.
+
+The main idea is to distribute this CLI as a [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) and the use experience should be similar to the `kubectl` CLI, as you can see in the example below:
+
+```bash
+kubectl prometheus-operator create prometheus --name my-prometheus --namespace my-namespace --replicas 2 --shards 2
+```
+
+### Where the code will live?
+
+The prometheus-operator-cli code will be placed in a dedicated repository under the [Prometheus organization](https://github.com/prometheus-operator), this will allow us to have a separeted lifecycle for the Prometheus Operator and the Prometheus Operator CLI and the following benefits:
+
+- We can have a dedicated team to maintain the Prometheus Operator CLI
+- We can use the Prometheus Operator client as a customer and be more careful with the changes that we make in the Prometheus Operator codebase
+- We can have a dedicated release cycle for the Prometheus Operator CLI
+
+### General features
+
+The Prometheus Operator CLI should have the following features, but not limited to:
+
+#### Resource scaffolding
+
+Allow users create Prometheus Operator resources using a simple CLI interface, the CLI should generate the resource manifest based on the user input.
```

Comment:
If someone wants to create a prometheus instance, are we scaffolding only the Prometheus resource or every other resource that is needed for it to run smoothly?

I don't see much value if it's only the Prometheus resource to be honest

Reply:
Agreed, I'd love to see an interactive CLI which would generate all things for you

````
$ kubectl prometheus-operator create prometheus example --interactive
How many replicas (default: 2)?
How many shards (default: 1)?
Create service account (Yes/No, default: Yes)?
Service account name (default: example)?
Create service (Yes/No, default: Yes)?
...
Create Pod disruption budget (Yes/No, default: Yes)?
Service monitor selector (default: {})?
Service monitor namespace selector (default: {})?
...
````

Reply:
I agree, this would be useful. Generating the RBAC, CRDs depending on the user's needs which will be understood from a questionaire like format seems cool.

Reply:
> If someone wants to create a prometheus instance, are we scaffolding only the Prometheus resource or every other resource that is needed for it to run smoothly?
> 
> I don't see much value if it's only the Prometheus resource to be honest

Both options are valid, you can have users with different use cases.
1. User deploying a entire stack: The interactive mode idea @simonpasquier shared is super cool and useful on this context, will help people deploy a entire stack easily.

3. User only creating extra resources for an existing stack, like we have with kubernetes create command, you can create a deployment without expose it and create a service independent of a deployment.

Reply:
> Agreed, I'd love to see an interactive CLI which would generate all things for you
> 
> ```
> $ kubectl prometheus-operator create prometheus example --interactive
> How many replicas (default: 2)?
> How many shards (default: 1)?
> Create service account (Yes/No, default: Yes)?
> Service account name (default: example)?
> Create service (Yes/No, default: Yes)?
> ...
> Create Pod disruption budget (Yes/No, default: Yes)?
> Service monitor selector (default: {})?
> Service monitor namespace selector (default: {})?
> ...
> ```

Just thinking out loud. 
Do you think we should make it a more generic prometheus cli and maybe in the future unite efforts with Thanos Operator?

Reply:
I think this is thinking too much about the future, the CLI here is still under design and will be experimental for quite a while, prom and thanos operator is also very much uncertain. Now creating a CLI for thanos, joining projects and joining CLIs is just so distant... no need to consider something like this at the moment 😅